pac4j-saml/src/main/java/org/pac4j/saml/store/HttpSessionStore.java

```diff
@@ -86,17 +89,27 @@ public void set(final String messageID, final XMLObject message) {
      */
     @Override
     public Optional<XMLObject> get(final String messageID) {
-        final LinkedHashMap<String, XMLObject> messages = getMessages();
-        final XMLObject o = messages.get(messageID);
+        final LinkedHashMap<String, String> messages = getMessages();
+        final String o = messages.get(messageID);
         if (o == null) {
-            log.debug("Message {} not found in session {}", messageID, context.getSessionStore().getOrCreateSessionId(context));
+            log.debug("Message {} not found in session {}",
+                    messageID, context.getSessionStore().getOrCreateSessionId(context));
             return Optional.empty();
         }
 
-        log.debug("Message {} found in session {}, clearing", messageID, context.getSessionStore().getOrCreateSessionId(context));
+        log.debug("Message {} found in session {}, clearing",
+                messageID, context.getSessionStore().getOrCreateSessionId(context));
         messages.clear();
         updateSession(messages);
-        return Optional.of(o);
+
+        try {
+            XMLObject message = XMLObjectSupport.unmarshallFromReader(
+                    Configuration.getParserPool(), new StringReader(o));
+            return Optional.of(message);
+        } catch (Exception e) {
+            log.error("Error unmarshalling message from input stream", e);
+            return Optional.empty();
+        }
```

Comment:
You should likely move these lines into a method next to one in Configuration that does the serialization.